### PR TITLE
[Core] Add runtime_env option to actor and task options, with conda_env

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Adds the option `runtime_env` to `.options()` for actors and tasks, and also to `JobConfig` (to be passed in via `ray.init()`).  For now, `runtime_env` is a `dict` which may contain a key-value pair `"conda_env": env` where `env` is a string which is the name of the conda environment the actor or task should run in.  Child actors and tasks will inherit the conda environment of the parent.

The user is responsible for ensuring the conda environment with the specified name already exists on all nodes in their Ray cluster.

Example:
```python
import ray
import tensorflow as tf

ray.init()

@ray.remote
def tf_version():
    return ("Tensorflow " + tf.__version__)

print(ray.get(tf_version.options(runtime_env={"conda_env": "tf1"}).remote()))
# Tensorflow 1.15.0
print(ray.get(tf_version.options(runtime_env={"conda_env": "tf2"}).remote()))
# Tensorflow 2.3.0
```

Or:
```python
runtime_env = {"conda_env": f"tf-{tf_version}"}
ray.init(job_config=JobConfig(runtime_env={"conda_env": "my_conda_env"}))
# All subsequent actors/tasks will be started in my_conda_env unless overridden in .options()
```

## Related issue number

RFC: https://github.com/ray-project/ray/issues/14019
Closes https://github.com/ray-project/ray/pull/14430

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_advanced_3.py [hunk 1]
```diff
@@ -785,6 +786,20 @@ def nested_get(self, key):
...
+    with pytest.raises(ValueError):
+        # Env tf2 should not exist.
+        env_dir = get_conda_env_dir("tf2")
+    tf2_dir = tmp_path / "tf2"
+    tf2_dir.mkdir()
+    env_dir = get_conda_env_dir("tf2")
+    assert (env_dir == str(tmp_path / "tf2"))
+    os.environ["CONDA_PREFIX"] = ""
+
+
```
Comment (edoakes) [inline python/ray/tests/test_advanced_3.py:802]: is it also possible to have an end-to-end test that creates two different conda envs and starts actors in each? we should also test the inheritance properties (child tasks/actors will be in same env)

**Review by edoakes — COMMENTED** [review]
Logic looks good. A couple of things:

- We need to have end-to-end tests for this.
- Need to update the docs (change the existing Serve docs and add a docs page for runtime_env).
- Can we also add runtime_env as an option to the `JobConfig` which will default it for tasks/actors submitted by the driver?

**Comment by edoakes** [comment]
@pcmoritz could you please take a look? This is the first part of the [runtime_env RFC](https://github.com/ray-project/ray/issues/14019).

To add to @architkulkarni's description above, it'll also be supported in the `JobConfig` to set it for all tasks/actors submitted by the driver.

**Comment by pcmoritz** [comment]
The `.options(runtime_env={"conda_env": ...})` API change LGTM!

**Comment by ericl** [comment]
cc @iycheng this is related to your current work

File: python/ray/tests/test_advanced_3.py (hunk 1 above)
Comment (architkulkarni) [reply python/ray/tests/test_advanced_3.py:802]: We wondered about this back when we did the Serve conda feature, but I don't think we were able to figure out how to do it--I'll continue to think about it.  For the underlying environment variable override feature, the inheritance properties are tested, but you're right that we should still have an end-to-end test for the conda env feature.

**Comment by architkulkarni** [comment]
Thanks for the review @edoakes!

What changes did you have in mind for the Serve docs?  Currently the Serve API and implementation for conda envs is unchanged, but maybe it should be changed:  should we add `runtime_env={"conda_env": ...}` as a field in `BackendConfig` in place of the current `env=CondaEnv(...)`, or should we skip `BackendConfig` and just tell the user to pass in the actual `runtime_env` to `ray_actor_options`?  Or something else?

The only Serve-specific conda feature is that if a conda env is unspecified, we automatically start backends in the conda env of the driver script, but this shouldn't affect the choice we make above.

**Comment by edoakes** [comment]
@architkulkarni ah yeah we should pivot to explicitly using the `runtime_env` dictionary so we don't need to replicate the interface. Actually, this should already work if it's passed in to `ray_actor_options` I believe.

**Comment by ericl** [comment]
Can we merge this?

**Comment by architkulkarni** [comment]
I'm still working on the end-to-end test which should be ready by the end of the day.

**Review by edoakes — APPROVED** [review]
LGTM. Thanks for getting the end-to-end tests set up, that will pay dividends.

**Comment by bklaubos** [comment]
@architkulkarni @edoakes 

> Adds the option runtime_env to .options() for actors and tasks, and also to JobConfig (to be passed in via ray.init())
Q: Can the JobConfig for runtime_env be passed via ray.connect() call?

**Comment by edoakes** [comment]
For future reference yes it can!

`ray.util.connect(job_config=job_config)`

**Pull request merged**